### ISSUE TITLE
Deletion of category freezes Qualcoder and bug in while loop

### ISSUE THREAD

**Issue by IngaIbs** [ccbogel/QualCoder#513]
Hi,
I have a project which froze when I opened some categories of the coding reports, namely coding comparison, coding comparison by file, code frequencies, view graph and code relations. It seems like the algorithm was stuck in while loops due to deleted categories. However, when I deleted categories in another project there was no problem, so I am not sure, why it is a problem here and I did not find out yet how to reproduce the problem. I found this out by outputting me the category it was stuck in, which was over and over some deleted category and looking at the log file. 

One example trace when trying to open the code relations was:
  ```
File "/usr/local/lib/python3.8/dist-packages/Qualcoder-2.5-py3.8.egg/qualcoder/__main__.py", line 1011, in report_code_relations
    ui = DialogReportRelations(self.app, self.ui.textEdit)

File "/usr/local/lib/python3.8/dist-packages/Qualcoder-2.5-py3.8.egg/qualcoder/report_relations.py", line 95, in __init__
    self.fill_tree()

 File "/usr/local/lib/python3.8/dist-packages/Qualcoder-2.5-py3.8.egg/qualcoder/report_relations.py", line 435, in fill_tree
    if item.text(1) == 'catid:' + str(c['supercatid']):
```

During the debug process I realized that the while loops that are designed to prevent the freeze are wrongly defined, since  the line `while len(cats) > 0 or count < 10000:` does not break if at least one of the two conditions is true. However I think it would be desired for it to break if one of the conditions is false so it should be changed to `while not (len(cats) < 0 or count > 10000)`.

When I changed these statements in
 
report_compare_coder_file.py , line 509
reports.py, lines 360 & 385
report_relations.py, lines  428

and changed `while supercatid is not None: ` to `while not (supercatid is None or count > 10000): ` in
view_graph_original.py, line 179

I was able to fix the problem. However this does not fix the underlying problem of old categories being in the queue.

Is use  
 - OS: Linux Mint 20.2
 Qualcoder 2.7
but the same problem has occured with the same project on my colleagues' linux distros.

Please let me know if I can be of any help or if you need further information. It works on my setup for now, but I wanted to let you know about the issue.

All the best,
Inga

**Comment by ccbogel** [comment]
will close this isssue soon if no responses.

**Comment by IngaIbs** [comment]
Hi sorry, I was away for a few days.
So I ran `SELECT catid, name, supercatid FROM code_cat` and two deleted categories were found that are designated to a non-existing catid.
However when I updated the catid as you proposed, it reintroduced the categories with their codes, though they were deleted. 

I deleted them again and tested if it worked this time by replacing a `while not (len(cats) < 0 or count > 10000):` with `while(len(cats) > 0` to see if it froze up again and it did. 

Otherwise I really like the qualcoder project, I did not have many problems so far!
